app/models/concerns/foreman_remote_execution/host_extensions.rb

```diff
@@ -3,16 +3,45 @@ module HostExtensions
     extend ActiveSupport::Concern
 
     included do
+      alias_method_chain :build_required_interfaces, :remote_execution
       has_many :targeting_hosts, :dependent => :destroy, :foreign_key => 'host_id'
+      validate :execution_interface_count
     end
 
-    # create or overwrite instance methods...
-    def instance_method_name
+    def execution_interface
+      get_interface_by_flag(:execution)
     end
 
-    module ClassMethods
-      # create or overwrite class methods...
-      def class_method_name
+    def remote_execution_proxy(provider)
+      candidate_proxies_for(provider).first
+    end
+
+    private
+
+    def build_required_interfaces_with_remote_execution(attrs = {})
+      build_required_interfaces_without_remote_execution(attrs)
+      self.primary_interface.execution = true if self.execution_interface.nil?
+    end
+
+    def candidate_proxies_for(provider)
+      [
+        if execution_interface && execution_interface.subnet
+          execution_interface.subnet.remote_execution_proxies.with_features(provider)
+        end,
+
+        if Setting[:remote_execution_fallback_proxy]
+          smart_proxies.with_features(provider)
+        end,
+
+        if Setting[:remote_execution_global_proxy]
+          ::SmartProxy.authorized.with_features(provider)
+        end
+      ].flatten.compact
+    end
+
+    def execution_interface_count
```

Comment:
this won't catch adding second execution interface through interfaces API, otoh validation on interface level can be tricky, it would have to check host's in memory interfaces as well